src/algebra/invertible.lean

```diff
@@ -177,7 +177,7 @@ by simp only [←two_mul, mul_inv_of_self]
 instance invertible_inv_of [has_one α] [has_mul α] {a : α} [invertible a] : invertible (⅟a) :=
 ⟨ a, mul_inv_of_self a, inv_of_mul_self a ⟩
 
-@[simp] lemma inv_of_inv_of [monoid α] {a : α} [invertible a] [invertible (⅟a)] :
+@[simp] lemma inv_of_inv_of [monoid α] {a : α} [invertible a] :
```

Comment:
I believe this is on purpose. You're effectively weakening the lemma with this change. Other lemmas underwent the opposite change recently, I'm thinking of `fintype` ones. @kmill will know which.

Even if the instances of `invertible` can't synthesize `invertible (⅟a)` in two ways (I checked), rewriting a goal *can* generate a non-defeq `invertible` instance.

Reply:
Yeah I wasn't 100% sure about this, I'll revert this one

Reply:
Yeah, `invertible` is a Type-valued subsingleton (like `fintype`), at least in this `monoid` context, so you want fully generalized instances for things that appear in the theorem statement.  [Here is one fintype example](https://github.com/leanprover-community/mathlib/pull/13055/files), where although every `finset` has a `fintype` instance, the theorem is more applicable if you generalize it -- no need for `convert` and friends.

Reply:
@kmill agreed! It would be great to have some more documentation, and maybe even a linter / metaprogram to help us follow this best practice :)

Reply:
Sounds like a you-job ;)

Reply:
Agree, it's always easiest to do when other people have agreed on what exactly the "rules" or guidelines for best practice are first though. Saves me doing things wrong and finding out the hard way 🤣

Reply:
Maybe this is the rule: "whenever there is an instance for a Type-valued class that, given all the other classes present, is a subsingleton, then that instance should be an instance argument."

I'm not sure how subsingleton determination should work -- maybe having the linter try to synthesize `subsingleton` instances is sufficient?